Add data_viz_in_OSS (#2)
arcus/education_modules: data_visualization_in_open_source_software/data_visualization_in_open_source_software.md

![A scatter plot, showing each patient's age along the x-axis and their glucose level on the y-axis.](media/intro_1.png)

Analysts frequently find themselves in situations where they want to depict more than two dimensions at a time. While it is possible to create 3D plots and animated plots, there are often simpler ways to express additional dimensions --- and simpler is nearly always better when it comes to visualizations!

You can use **color** to communicate either continuous (e.g. darker = higher value) or categorical values (e.g. red = "no" and blue = "yes"). **Shape** (e.g. circle, triangle, cross) or **line type** (e.g. solid or dashed) can also indicate category membership and have the advantage of not relying on end users' ability to distinguish color. **Size** is often used to indicate count or population values.

![A scatter plot of age vs glucose with lines of best fit, with patients in Class 1 shown as yellow dots with a solid best fit line, and Class 2 as blue triangles with a blue dashed line.](media/intro_2.png)

> 

Reviewer: (if adding here, presumably also adding to others)

> 

Reviewer 2: Yes, thanks. I think for now I'll leave it and hope that the existing explanation in the text is sufficient, but I'll keep an eye on this as potential enhancement down the line.

Another valuable tool is **faceting**. This is when similar plots (that share one or both axes) are lined up next to each other to facilitate comparison. For example, when you want to show the relationship between two variables, X and Y, for several different subsets of the data, you can generate a scatter plot of X and Y for each subset of the data as its own facet. If you have more than two or three groups, faceting generally produces clearer plots than using color or shape to distinguish groups.

> ![Two scatter plots side by side showing age vs glucose levels for patients in Class 1 (left) and Class 2 (right). The y-axis is aligned to facilitate comparison across groups.](media/intro_3.png)

Reviewer: ```suggestion
![The same data from above split into two separate scatter plots side by side, one for patients in Class 1 and one for Class 2. The y-axis is aligned to facilitate comparison across groups.](media/intro_3.png)
```

Most of these plot elements can be applied across a wide range of kinds of visualizations. Knowing how to identify these key plot elements will help you select effective visualizations for your own data.

Before we dive into how plotting works in practice, though, we'll explore some popular open source tools for data visualization.

## Plotting libraries in R and python

There are many options available for creating data visualizations in R and python. For this module, we'll focus on just one plotting library in each: `ggplot2` in R, and `seaborn` in python.

### `ggplot2` in R

> R comes with plenty of plotting functions ready to go in base R, and for quick-and-dirty visualizations, that is usually the fastest option. The popular library [`ggplot2`](https://ggplot2.tidyverse.org/) provides much more flexibility and control over your visualizations, however, and if you're creating visualizations to share (e.g. in a manuscript or presentation), you'll probably find you need to switch to `ggplot2` at some point anyway to get more control over the appearance of your visualizations.

Reviewer: since this is slated as for beginners, I wonder if "base R" is a term that should be defined/clarified?

To get started with `ggplot2` in R on your own machine, you'll need to install it, and then load the library in R:

```r
# You only need to install it once
install.packages("ggplot2")

> # You'll need to load the library anew for each R session

Reviewer: ```suggestion
# You'll need to load the library again for each R session
```
not particularly common word, not sure if it may introduce language barriers for some

library("ggplot2")
```

The `ggplot2` package is part of the [tidyverse](https://www.tidyverse.org/) of R packages, and is designed to work well with other tidyverse libraries like dplyr and tidyr, all of which work together to streamline common data science workflows in R. It's also fine to use `ggplot2` on its own, though.

For an excellent quick reference, see the [`ggplot2` cheatsheet](https://ggplot2.tidyverse.org/#cheatsheet). It includes a tremendous amount of information in a very compact format, so it's not great for people just getting started with `ggplot2`, but it's a valuable reference to keep on hand for when you start making plots for your own analyses.

### `seaborn` in python

In python, there's a similar situation: The [`seaborn`](https://seaborn.pydata.org/) library is a plotting library that works on top of the more basic and powerful library [matplotlib](https://matplotlib.org/), making it easier to use for common data science applications.

> To use `seaborn` in python on your own machine, you'll need to [install `seaborn`](https://seaborn.pydata.org/installing.html) it first. You can use either PyPi (`pip install seaborn`) or Anaconda (`conda install seaborn`) to install it, whichever you use for your other python modules.

Reviewer: ```suggestion
To use `seaborn` in python on your own machine, you'll need to [install `seaborn`](https://seaborn.pydata.org/installing.html) first. You can use either PyPi (`pip install seaborn`) or Anaconda (`conda install seaborn`) to install it, whichever you use for your other python modules.
```

- Although planning a plot carefully before you start is a good idea and will save you time, don't expect that it will save you from having to re-work it. Creating a good visualization is almost always an iterative process --- that's one reason it's such an advantage to do your plotting in a scripted language like R or python! If you want to make a small change to a plot, you don't have to start from scratch, you can just tweak the code and re-run it.
</div>
****

Which of the following are good strategies to extend the accessibility of your data visualizations? Select all that apply.
 [[X]] Avoid using color as the sole means of distinguishing important information
 [[X]] Provide informative alt text (for online visuals) or captions
 [[X]] Avoid cluttered, complex visuals
 [[X]] Present the information from the plot in alternate forms (a data table in an appendix, narrative description in the text, etc.)

>  ***

Reviewer: ```suggestion
 ****
```

<div class = "answer">
All of these are good strategies!

In general, review your work asking yourself "If the reader didn't have visual access, would this still communicate everything I intend it to?" It may not always be possible to make your work maximally accessible, but there's often a lot you can do to improve it.

**Note**: We've discussed visual accessibility exclusively here because that is a common stumbling block for data visualizations, but access is much more than vision. For broader guidance on creating accessible content, see the [Web Content Accessibiility Guidelines](https://www.w3.org/WAI/WCAG21/quickref/) and [The Inclusive Learning Design Handbook](https://handbook.floeproject.org/about/).
 </div>

>  ***

Reviewer: ```suggestion
 ****
```

this quiz q isn't rendering properly -- I think it's just the number of asterisks?

comment:   Introduction to principles of data vizualization and typical data vizualization workflows using two common open source libraries: ggplot2 and seaborn.

long_description: This module introduces ggplot2 and seaborn, popular data visualization libraries in R and python, respectively. It lays the groundwork for using ggplot2 and seaborn by 1) highlighting common features of plots that can be manipulated in plot code, 2) discussing a typical data visualization workflow and best practices, and 3) discussing data preparation for plotting. This content will be most useful for people who have some experience creating data visualizations and/or reading plots presented in research articles or similar contexts. Some prior exposure to R and/or python is helpful but not required. This is appropriate for beginners.

estimated_time: 20 minutes

@learning_objectives

After completion of this module, learners will be able to:

* identify key elements in a plot that communicate information about the data
* describe the role ggplot2 and seaborn play in the R and python programming languages, respectively
* describe a typical data vizualization workflow

> * recognize how data structure may impact the creation of data vizualizations

Reviewer 2: Yeah, I agree. This is something that feels important to me (which is why it's worded so grandly in the LOs), but discussing it satisfactorily without getting into any coding examples is a little tricky, so it's not quite landing. I'll have a think about strengthening that section vs. nerfing the LO.

> * recognize how data structure may impact the creation of data vizualizations

Reviewer 2: I opted to nerf the LO ;) That content deserves a more thorough treatment, but in a module that can get into code examples, I think (so not this one).

[[?]] Hint: See [data visualization workflow](#data-visualization-workflow) for a review.
****
<div class = "answer">
Although the first two answers seem plausible, they don't really describe a typical workflow:

- Everyone, including experts, reuses code from example plots and previous visualizations (that's why [galleries of example plots](#data-visualization-workflow) are so popular!). You shouldn't expect to be able to write `ggplot2` or `seaborn` code from a blank slate. As with so many programming skills, the main difference between a `ggplot2`/`seaborn` expert and a novice is that the expert knows exactly what to google!
- Although planning a plot carefully before you start is a good idea and will save you time, don't expect that it will save you from having to re-work it. Creating a good visualization is almost always an iterative process --- that's one reason it's such an advantage to do your plotting in a scripted language like R or python! If you want to make a small change to a plot, you don't have to start from scratch, you can just tweak the code and re-run it.

> </div>

Reviewer 2: Yeah, I hear that. What I'm trying to do with this is draw their attention to some of the nuance they may have missed in the workflow description, especially the fact that even experts don't expect to be able to write perfect plot code on the first try without referencing examples etc. But I'll rethink this question completely and see if I can get at that from another angle entirely.

> </div>

Reviewer 2: I feel like I got a non-sneaky question to replace this one that still does the work I was hoping this would do re drawing learners' attention to some of the nuance of the workflow section. Let me know what you think!

> In python, there's a similar situation: The [`seaborn`](https://seaborn.pydata.org/) library is a plotting library that works on top of the more basic and powerful library [matplotlib](https://matplotlib.org/), making it easier to use for common data science applications.

Reviewer 2: Good catch, thanks!

</div>
****

For someone who has spent a lot of time getting good at `ggplot2` or `seaborn`, which of the following would likely be part of their typical workflow when building a new data visualization? Select all that apply.

[[X]] Spending time thinking about (and maybe drawing) the plot they want before they start coding
[[X]] Searching for example code for similar plots online
[[X]] Going through several different versions of their code before they get a plot that they want to keep
****
<div class = "answer">
All of these are typical for experienced programmers!

> Sometimes it can feel discouraging when, as a beginner, you feel like everything you code takes several false starts before you make any real progress. The truth is that writing code is full of false starts, even for the most expert programmers. Everyone, including experts, reuses code from example plots and previous visualizations (that's why [galleries of example plots](#data-visualization-workflow) are so popular!). You shouldn't expect to be able to write `ggplot2` or `seaborn` code from a blank slate. As with so many programming skills, the main difference between a `ggplot2`/`seaborn` expert and a novice is that the expert knows exactly what to google!

Reviewer: I think the reworking of this question makes a lot more sense, and I super love how it also turned into an opportunity for normalization of the sometimes frustrating and iterative process that is learning how to code :)
